src/analysis/confusion_matrix.rs

```diff
@@ -0,0 +1,182 @@
+//! Module to compute the confusion matrix of a set of predictions.
+
+use std::hash::Hash;
+use std::collections::HashMap;
+use linalg::Matrix;
+
+/// Returns a matrix C where C_ij is the count of the samples which were
+/// predicted to lie in class j but actually lie in class i. (i.e.
+/// the row is the true class and the column is the predicted.).
+///
+/// # Arguments
+/// * `predictions` - A series of model predictions.
+/// * `targets` - A slice of equal length to predictions, containing the target results.
+/// * `labels` - If not provided then the returned matrix rows and columns correspond to
+///              the distinct elements of union(predictions, targets), in increasing
+///              order. If labels are provided then these are used instead, in the order
+///              provided. Note that this means that the confusion matrix will only contain
+///              entries for the elements of labels.
+///
+/// # Examples
+/// ```
+/// use rusty_machine::analysis::confusion_matrix::confusion_matrix;
+/// use rusty_machine::linalg::Matrix;
+///
+/// let truth       = vec![2, 0, 2, 2, 0, 1];
+/// let predictions = vec![0, 0, 2, 2, 0, 2];
+///
+/// let confusion = confusion_matrix(&predictions, &truth, None);
+///
+/// let expected = Matrix::new(3, 3, vec![
```

Comment:
We could also use the `matrix!` macro here - though I think it would be better to modify doc-facing code all at once.